// actionhero/actions/file.js
'use strict'
const fs = require('fs')
const path = require('path')
const {promisify} = require('util')
const {Action} = require('./../index.js')

module.exports = class File extends Action {
  constructor () {
    super()
    this.name = 'file'
    this.description = 'I send a file from an action'
    this.inputs = {
      type: {
        required: true,
        default: 'file',
        validator: (p) => {
          if (['file', 'string', 'buffer'].indexOf(p) < 0) {
            throw new Error('type should be file or string')
          }
          return true
        }
      }
    }
  }

  async run (data) {
    data.toRender = false

    if (data.params.type === 'file') {
      return data.connection.sendFile('simple.html')
    }

    const readFile = promisify(fs.readFile)
    const buffer = await readFile(path.join(__dirname, '..', 'public', 'simple.html'))

    if (data.params.type === 'buffer') {
      return data.connection.sendFile(buffer)
    }

    if (data.params.type === 'string') {
      return data.connection.sendFile(buffer.toString())
    }
  }
}
